File: 2_examples/java/src/main/java/no/systek/graphqlworkshop/resolvers/OrderMutationResolver.java
package no.systek.graphqlworkshop.resolvers;

import com.coxautodev.graphql.tools.GraphQLMutationResolver;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;
import no.systek.graphqlworkshop.storage.DataSource;
import no.systek.graphqlworkshop.storage.Order;
import org.springframework.stereotype.Component;

@Component
public class OrderMutationResolver implements GraphQLMutationResolver {

  private final DataSource dataSource;

  public OrderMutationResolver(DataSource dataSource) {
    this.dataSource = dataSource;
  }

  public Order order(List<NewDish> dishes) {
    Order order = new Order(
        new Double((Math.random() * Long.MAX_VALUE)).longValue(),
        LocalDateTime.now(),
        dishes.stream()
            .map(dish -> dataSource.getDish(dish.getDishId()))
            .collect(Collectors.toList())
    );
    dataSource.addOrder(order);
    return order;
  }
}

File: 3_examples/kotlin/src/main/kotlin/no/systek/graphqlworkshop/resolvers/OrderMutationResolver.kt
package no.systek.graphqlworkshop.resolvers

import com.coxautodev.graphql.tools.GraphQLMutationResolver
import no.systek.graphqlworkshop.storage.DataSource
import no.systek.graphqlworkshop.storage.Dish
import no.systek.graphqlworkshop.storage.Order
import org.springframework.stereotype.Component
import java.time.LocalDateTime

@Component
class OrderMutationResolver(
    private val dataSource: DataSource
) : GraphQLMutationResolver {
    fun order(dishes: Collection<NewOrder>): Order {
        val order = Order(
            LocalDateTime.now(),
            dishes.toDishes())
        dataSource.addOrder(order)
        return order
    }

    fun Collection<NewOrder>.toDishes(): List<Dish> = this.map { it.toDish() }

    fun NewOrder.toDish(): Dish = dataSource.getDish(this.dishId)
}

data class NewOrder(
    val dishId: Long,
    val quantity: Long
)
